### PR TITLE
Migrate C# client-side reducer enum to the new syntax

### PR DESCRIPTION
**Pull request by RReverser**
# Description of Changes

As per https://github.com/clockworklabs/SpacetimeDBPrivate/pull/1198.

This makes reducers use the same codegen structure as tables and other tuple types, but nests them inside a `Reducer` class as subclasses for pattern-matching, like we allow on tagged enums.

The reason we don't use a tagged enum here anymore is because we don't need BSATN sum type support for the Reducer enum, and it imposes unnecessary restrictions like max 256 elements.

While testing the changes, simplified the C# CI a bit by reusing the `tools~/write-nuget-config.sh` script from the C# SDK repo instead of duplicating its logic.

# API and ABI breaking changes

API breaking change for the client.

# Expected complexity level and risk

2 - a breaking change, but shouldn't affect too much client code.

# Testing

Waiting for CI to test it against client SDK, it's a bit quicker than manual.

- [x] *Write a test you've completed here.*
- [ ] *Write a test you want a reviewer to do here, so they can check it off when they're satisfied.*

### CODE REVIEW

File: .github/workflows/ci.yml
Comment (gefjon) [inline .github/workflows/ci.yml:123]: Why did this file change? Is that related to this PR?

File: crates/bindings-csharp/Runtime/Internal/Autogen/RawRowLevelSecurityDefV9.cs
Comment (gefjon) [inline crates/bindings-csharp/Runtime/Internal/Autogen/RawRowLevelSecurityDefV9.cs:21]: Do these files get auto-formatted? If so, why did they change?

**Review by gefjon — APPROVED** [review]
I have no way to test this, but it appears reasonable. I'd like clarity on the CI changes before you merge, but otherwise this looks good to go.

File: .github/workflows/ci.yml
Comment (RReverser) [reply .github/workflows/ci.yml:123]: I was trying to test this and wanted to regenerate client SDK as part of the CI. That backfired - cross-repo testing is still pretty damn hard - so the test will have to be as part of client SDK PR, so I reverted the `spacetime generate` changes but left this one as it's purely for clarity. I'm happy to split it out as it's not related anymore.

File: crates/bindings-csharp/Runtime/Internal/Autogen/RawRowLevelSecurityDefV9.cs
Comment (RReverser) [reply crates/bindings-csharp/Runtime/Internal/Autogen/RawRowLevelSecurityDefV9.cs:21]: No, I changed the C# codegen logic while unifying reducer, tables and type generation. They're not autoformatted.

File: .github/workflows/ci.yml
Comment (gefjon) [reply .github/workflows/ci.yml:123]: I'm fine to keep it in as long as you add a note about it to the PR description.

Comment (RReverser) [reply .github/workflows/ci.yml:123]: Done

**Review by bfops — APPROVED** [review]
Approving CLI changes as a codeowner. I didn't really review; the changes are in C#-specific code that probably should be in an un-owned file anyway.

**Pull request merged**
